### PR TITLE
Security Advisories

### PR DESCRIPTION
**Pull request by glaubinix**
This pull request integrates security advisories into packagist.org. It is built in a way to additional sources can be integrated.

**Users are able to:**
* see how many advisories are available for every package
* see which versions of a package are affected by one or more advisory
* see detailed information about advisories for a package
* query the data via an API

**Adjustments to the package view page:**
* number of security advisories
* red icon in the version list to see which versions are affected
Screenshot 2019-11-23 at 13 22 04

**New package security advisory page**
* list all advisories for a package
* via the query param `version` matching advisories will be highlighted
Screenshot 2019-11-23 at 13 24 56

### CODE REVIEW

File: src/Packagist/WebBundle/Controller/PackageController.php [hunk 1]
```diff
@@ -1120,6 +1138,45 @@ public function versionStatsAction(Request $req, Package $package, $version)
...
+            $version = $this->getDoctrine()->getRepository(Version::class)->findOneBy([
+                'name' => $name,
+                'id' => $versionId,
+            ]);
+            if ($version) {
+                foreach ($securityAdvisories as $advisory) {
+                    $versionParser = new VersionParser();
+                    $affectedVersionConstraint = $versionParser->parseConstraints($advisory['affectedVersions']);
+                    if (!isset($data['hasVersionSecurityAdvisories'][$version->getId()]) && $affectedVersionConstraint->matches(new Constraint('=', $version->getNormalizedVersion()))) {
+                        $data['matchingAdvisories'] = $advisory['id'];
```
Comment (Seldaek) [inline src/Packagist/WebBundle/Controller/PackageController.php]: This should be an array of ids matching I believe?

File: src/Packagist/WebBundle/Controller/PackageController.php [hunk 2]
```diff
@@ -1120,6 +1138,45 @@ public function versionStatsAction(Request $req, Package $package, $version)
...
+
+        $data['matchingAdvisories'] = [];
+        if ($versionId = $request->query->getInt('version')) {
+            $version = $this->getDoctrine()->getRepository(Version::class)->findOneBy([
+                'name' => $name,
+                'id' => $versionId,
+            ]);
+            if ($version) {
+                foreach ($securityAdvisories as $advisory) {
+                    $versionParser = new VersionParser();
```
Comment (Seldaek) [inline src/Packagist/WebBundle/Controller/PackageController.php]: No need to create one per iteration?

File: src/Packagist/WebBundle/Controller/PackageController.php [hunk 3]
```diff
@@ -1120,6 +1138,45 @@ public function versionStatsAction(Request $req, Package $package, $version)
...
+                    $affectedVersionConstraint = $versionParser->parseConstraints($advisory['affectedVersions']);
+                    if (!isset($data['hasVersionSecurityAdvisories'][$version->getId()]) && $affectedVersionConstraint->matches(new Constraint('=', $version->getNormalizedVersion()))) {
```
Comment (Seldaek) [inline src/Packagist/WebBundle/Controller/PackageController.php]: The isset here seems off. Seems it was copied over from viewPackageAction

File: src/Packagist/WebBundle/Controller/PackageController.php [hunk 4]
```diff
@@ -1120,6 +1138,45 @@ public function versionStatsAction(Request $req, Package $package, $version)
...
+                        $data['matchingAdvisories'] = $advisory['id'];
+                    }
+                }
+            }
+        }
```
Comment (Seldaek) [inline src/Packagist/WebBundle/Controller/PackageController.php]: Generally speaking once this block executes, the paginator will be in a bad state I assume, as this filters the list of securityAdvisories which are displayed? Might have missed something, but it sounds to me like the filtering should happen before paginator and count etc are assigned to $data?

File: src/Packagist/WebBundle/Controller/ApiController.php [hunk 5]
```diff
@@ -246,6 +247,32 @@ public function trackDownloadsAction(Request $request)
...
+            return trim($packageName);
+        }, explode(',', $request->get('packages')))));
+        $updatedSince = $request->query->getInt('updatedSince', 0);
+
+        /** @var array[] $advisories */
+        $advisories = $this->getDoctrine()->getRepository(SecurityAdvisory::class)->searchSecurityAdvisories($packageNames, $updatedSince);
+
+        $response = [];
+        foreach ($advisories as $advisory) {
+            $response[$advisory['packageName']][] = $advisory;
```
Comment (Seldaek) [inline src/Packagist/WebBundle/Controller/ApiController.php]: Should be returned nested as `{"advisories": {...}}`

File: src/Packagist/WebBundle/Controller/PackageController.php (hunk 4 above)
Comment (glaubinix) [reply src/Packagist/WebBundle/Controller/PackageController.php]: The pagination is currently unused. Removed it completely.

**Comment by Seldaek** [comment]
Thanks!

**Comment by scheb** [comment]
Really like this feature! Did you consider making use of this and let Composer warn about security advisories when you're using an affected version?

Recently, I had to deal with a security issue and I found it hard to let people know, because I don't know all of the users and there is no direct communication channel to all of them.

I know there are solutions out there to check your project for security issues, but these require addtional software to be added to your project/build - not everyone is doing this. Composer is the one piece of software that everyone uses to install their dependencies and every once in a while you do a composer install/update. I think it would be really nice if it would warn you when there is a security risk in your current dependencies. Would give these issues more visibility and hopefully encourage people to upgrade.

**Pull request merged**